src/Npgsql/TypeHandlers/CompositeHandlers/CompositeConstructorHandler`.cs

```diff
@@ -0,0 +1,67 @@
+﻿using System;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Threading.Tasks;
+using Npgsql.PostgresTypes;
+
+namespace Npgsql.TypeHandlers.CompositeHandlers
+{
+    sealed class CompositeConstructorHandler<TComposite, T1, T2, T3, T4, T5, T6, T7, T8> : CompositeConstructorHandler<TComposite>
+    {
+        delegate TComposite CompositeConstructor(in Arguments args);
+
+        readonly CompositeConstructor _constructor;
+
+        public CompositeConstructorHandler(PostgresType postgresType, ConstructorInfo constructorInfo, CompositeParameterHandler[] parameterHandlers)
+            : base(postgresType, constructorInfo, parameterHandlers)
+        {
+            var parameter = Expression.Parameter(typeof(Arguments).MakeByRefType());
+            var fields = Enumerable
+                .Range(1, parameterHandlers.Length)
+                .Select(i => Expression.Field(parameter, "Argument" + i));
+
+            _constructor = Expression
+                .Lambda<CompositeConstructor>(Expression.New(constructorInfo, fields), parameter)
+                .Compile();
+        }
+
+        public override async ValueTask<TComposite> Read(NpgsqlReadBuffer buffer, bool async)
+        {
+            await buffer.Ensure(sizeof(int), async);
+
+            var fieldCount = buffer.ReadInt32();
+            if (fieldCount != Handlers.Length)
+                throw new InvalidOperationException($"pg_attributes contains {Handlers.Length} fields for type {PostgresType.DisplayName}, but {fieldCount} fields were received.");
+
+            var args = default(Arguments);
+
+            foreach (var handler in Handlers)
```

Comment:
nit: add curlies

Reply:
There are a switch block with curlies below. Don't think that adding more will make code cleaner.

Reply:
I prefer to add curlies the moment the body is more than one line.